File: src/data/o1/Text_Justification/Text_Justification.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> result;
        int n = words.size();
        int index = 0;
        
        while(index < n){
            int totalChars = words[index].size();
            int last = index + 1;
            while(last < n){
                if(totalChars + 1 + words[last].size() > maxWidth) break;
                totalChars += 1 + words[last].size();
                last++;
            }
            
            int numberOfWords = last - index;
            int numberOfSpaces = maxWidth - totalChars + (numberOfWords -1);
            string line;
            
            // If we're on the last line or the line contains only one word, left-justify
            if(last == n || numberOfWords == 1){
                for(int i = index; i < last; i++){
                    line += words[i];
                    if(i != last -1){
                        line += ' ';
                    }
                }
                // Fill the remaining spaces
                while(line.size() < maxWidth){
                    line += ' ';
                }
            }
            else{
                // Middle justified
                int spaces = (maxWidth - (totalChars - (numberOfWords -1))) / (numberOfWords -1);
                int extra = (maxWidth - (totalChars - (numberOfWords -1))) % (numberOfWords -1);
                
                for(int i = index; i < last; i++){
                    line += words[i];
                    if(i != last -1){
                        line += string(spaces + (extra > 0 ? 1 : 0), ' ');
                        if(extra > 0) extra--;
                    }
                }
            }
            
            result.push_back(line);
            index = last;
        }
        
        return result;
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple case with exact fit
    std::vector<std::string> words = {"This", "is", "an", "example", "of", "text", "justification."};
    int maxWidth = 16;
    std::vector<std::string> expected_result = {
        "This    is    an",
        "example  of text",
        "justification.  "
    };
    assert(solution.fullJustify(words, maxWidth) == expected_result && "Test case 1 failed.");

    // Test case 2: Single word that exactly matches maxWidth
    words = {"Justify"};
    maxWidth = 7;
    expected_result = {"Justify"};
    assert(solution.fullJustify(words, maxWidth) == expected_result && "Test case 2 failed.");

    // Test case 3: Single word shorter than maxWidth
    words = {"Hello"};
    maxWidth = 10;
    expected_result = {"Hello     "};
    assert(solution.fullJustify(words, maxWidth) == expected_result && "Test case 3 failed.");

    // Test case 4: Multiple words, last line left-justified
    words = {"What", "must", "be", "acknowledgment", "shall", "be"};
    maxWidth = 16;
    expected_result = {
        "What   must   be",
        "acknowledgment  ",
        "shall be        "
    };
    assert(solution.fullJustify(words, maxWidth) == expected_result && "Test case 4 failed.");

    // Test case 5: Uneven space distribution
    words = {"Science", "is", "what", "we", "understand", "well", "enough", "to", "explain"};
    maxWidth = 20;
    expected_result = {
        "Science  is  what we",
        "understand      well",
        "enough to explain   "
    };
    assert(solution.fullJustify(words, maxWidth) == expected_result && "Test case 5 failed.");

    // Test case 6: Only one word per line, lines longer than words
    words = {"Single", "word", "per", "line"};
    maxWidth = 10;
    expected_result = {"Single    ", "word   per", "line      "};
    assert(solution.fullJustify(words, maxWidth) == expected_result && "Test case 6 failed.");

    // Test case 7: All words fit into one line, exact width
    words = {"One", "line", "test"};
    maxWidth = 12;
    expected_result = {"One     line", "test        "};
    assert(solution.fullJustify(words, maxWidth) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 655360; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/o1/Text_Justification/Text_Justification.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> result = new ArrayList<>();
        int index = 0;
        int n = words.length;
        
        while (index < n) {
            int totalChars = words[index].length();
            int last = index + 1;
            // Determine how many words can fit into the current line
            while (last < n) {
                if (totalChars + 1 + words[last].length() > maxWidth) break;
                totalChars += 1 + words[last].length();
                last++;
            }
            
            int numberOfWords = last - index;
            StringBuilder sb = new StringBuilder();
            // If it's the last line or the line contains only one word, left-justify
            if (last == n || numberOfWords == 1) {
                for (int i = index; i < last; i++) {
                    sb.append(words[i]);
                    if (i != last - 1) {
                        sb.append(" ");
                    }
                }
                // Fill the remaining spaces
                while (sb.length() < maxWidth) {
                    sb.append(" ");
                }
            } else {
                // Calculate the number of spaces
                int totalSpaces = maxWidth - (totalChars - (numberOfWords - 1));
                int spaceBetweenWords = totalSpaces / (numberOfWords - 1);
                int extraSpaces = totalSpaces % (numberOfWords - 1);
                
                for (int i = index; i < last; i++) {
                    sb.append(words[i]);
                    if (i != last - 1) {
                        for (int s = 0; s < spaceBetweenWords; s++) {
                            sb.append(" ");
                        }
                        if (extraSpaces > 0) {
                            sb.append(" ");
                            extraSpaces--;
                        }
                    }
                }
            }
            
            result.add(sb.toString());
            index = last;
        }
        
        return result;
    }
}

public class Text_Justification {
public static void runTests(Solution solution) {
    // Test case 1: Simple case with exact fit
    String[] words1 = {"This", "is", "an", "example", "of", "text", "justification."};
    int maxWidth1 = 16;
    List<String> expected_result1 = Arrays.asList(
        "This    is    an",
        "example  of text",
        "justification.  "
    );
    assert solution.fullJustify(words1, maxWidth1).equals(expected_result1) : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Single word that exactly matches maxWidth
    String[] words2 = {"Justify"};
    int maxWidth2 = 7;
    List<String> expected_result2 = Arrays.asList("Justify");
    assert solution.fullJustify(words2, maxWidth2).equals(expected_result2) : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: Single word shorter than maxWidth
    String[] words3 = {"Hello"};
    int maxWidth3 = 10;
    List<String> expected_result3 = Arrays.asList("Hello     ");
    assert solution.fullJustify(words3, maxWidth3).equals(expected_result3) : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Multiple words, last line left-justified
    String[] words4 = {"What", "must", "be", "acknowledgment", "shall", "be"};
    int maxWidth4 = 16;
    List<String> expected_result4 = Arrays.asList(
        "What   must   be",
        "acknowledgment  ",
        "shall be        "
    );
    assert solution.fullJustify(words4, maxWidth4).equals(expected_result4) : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Uneven space distribution
    String[] words5 = {"Science", "is", "what", "we", "understand", "well", "enough", "to", "explain"};
    int maxWidth5 = 20;
    List<String> expected_result5 = Arrays.asList(
        "Science  is  what we",
        "understand      well",
        "enough to explain   "
    );
    assert solution.fullJustify(words5, maxWidth5).equals(expected_result5) : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Only one word per line, lines longer than words
    String[] words6 = {"Single", "word", "per", "line"};
    int maxWidth6 = 10;
    List<String> expected_result6 = Arrays.asList("Single    ", "word   per", "line      ");
    assert solution.fullJustify(words6, maxWidth6).equals(expected_result6) : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: All words fit into one line, exact width
    String[] words7 = {"One", "line", "test"};
    int maxWidth7 = 12;
    List<String> expected_result7 = Arrays.asList("One     line", "test        ");
    assert solution.fullJustify(words7, maxWidth7).equals(expected_result7) : "Test case 7 failed. Expected " + expected_result7;
}


    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 5000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    